Frameworks/UIKit/NSTextContainer.mm

```diff
@@ -75,25 +119,69 @@ - (CGRect)lineFragmentRectForProposedRect:(CGRect)proposed
                             remainingRect:(CGRect*)remainingRect {
     CGRect totalRect = CGRectMake(0, 0, _size.width, _size.height);
     CGRect ret = CGRectIntersection(proposed, totalRect);
+    if ([_exclusionPaths count] == 0L || ret.size.width <= 0) {
+        return ret;
+    }
+
+    CGFloat padding = self.lineFragmentPadding;
+    if (padding <= 0.0) {
+        // If padding is invalid we will go into an infinite loop, use a default
+        padding = c_defaultPadding;
+    }
+
+    if (remainingRect) {
+        *remainingRect = CGRectZero;
+    }
 
-    if (_exclusionPaths != nil) {
-        for (UIBezierPath* path in(NSArray*)_exclusionPaths) {
-            CGRect shapeRect = _CGPathFitRect(path.CGPath, ret, _size, self.lineFragmentPadding);
-            ret = CGRectIntersection(shapeRect, ret);
+    std::vector<CGRect> exclusionRectsForHorizontal{};
+    for (size_t i = 0; i < _exclusionPathBoundingRects.size(); ++i) {
+        // If our proposed area doesn't intersect the bounding box of the exclusion zone at all, no need to compare against it
+        if (proposed.origin.y < _exclusionPathBoundingRects[i].origin.y + _exclusionPathBoundingRects[i].size.height &&
+            proposed.origin.y + proposed.size.height > _exclusionPathBoundingRects[i].origin.y &&
+            proposed.origin.x < _exclusionPathBoundingRects[i].origin.x + _exclusionPathBoundingRects[i].size.width) {
+            CGRect lineIntersection{ { _exclusionPathBoundingRects[i].origin.x, proposed.origin.y },
+                                     { _exclusionPathBoundingRects[i].size.width, proposed.size.height } };
```

Comment:
unfortunately clang format is really opinionated about there not being a newline here :/ Added a comment for readability